docs/manual/common/concepts/InternalAndExternalCommunication.md

```diff
@@ -0,0 +1,7 @@
+#Internal and external communication
+
+Lagom promotes use of asynchronous communication without preventing use of synchronous communication where necessary. For example, as mentioned in [[Polyglot systems with Lagom|PolyglotSystems]], legacy application components might use synchronous communication that cannot be changed. Within your system, services might communicate with other services in the same node and services in other nodes.  
```

Comment:
Here's my attempt at some more content for this page:

-----

As discussed in [[Lagom design philosophy|LagomDesignPhilosophy]], services should be _isolated_ and _autonomous_. They communicate by sending messages to each other over a network. At the same time, performance and resilience requirements often create the need to run each service on multiple nodes. The nodes of a single system also need to communicate over a network. While many principles are similar, these two types of communication have very different needs, and Lagom supports multiple options for implementing these different communication patterns.

* Different services in a system must communicate using loosely-coupled protocols and message formats. Coordinating change between different services can be difficult and costly. Isolation and autonomy are paramount in inter-service communication. For example:
  * [[Service calls|ServiceDescriptors]], either synchronous or asynchronous (streaming), allow services to communicate with each other using published APIs and standard protocols (HTTP and WebSockets).
  * Sending messages via a [[broker|MessageBroker]] such as Apache Kafka decouples communication even further.
* Nodes of a single service (collectively called a **cluster**) require less decoupling. They share the same code and are managed together, as a set, by a single team or individual. Intra-service communication can take advantage of mechanisms that have less overhead and better performance. For example:
  * [Akka remoting](http://doc.akka.io/docs/akka/2.4/general/remoting.html) is used internally by many Lagom components, and can also be used directly.
  * [[Distributed publish-subscribe|PubSub]] can be used for low-latency, at-most-once messaging between nodes.
  * Databases and other persistent storage can be seen as another way that nodes of a service communicate.

This diagram shows how three services, each distributed across three servers, might communicate internally and externally:

![lagom communication 001](https://cloud.githubusercontent.com/assets/44385/24385961/b3d50ae4-13b8-11e7-8dce-9debe9103b10.png)

Lagom promotes use of asynchronous communication without preventing use of synchronous communication where necessary. For example, as mentioned in [[Polyglot systems with Lagom|PolyglotSystems]], legacy application components might use synchronous communication that cannot be changed.

-----

I uploaded the image into the UX-Lagom Google Drive folder, as well as the original Keynote file I used to create it (Mac only, I'm afraid) and a PowerPoint export (which usually gets pretty mangled in translation, but might still be a useful starting point). Let me know if there are any adjustments you'd like me to make. I can do it directly in Keynote and export another PNG file. Alternatively, if we want to get Rich to redo it entirely along with others, that's fine too.

We might want a second image and paragraph to illustrate the internal system vs. external world boundary as well. We could adapt it from my comment on the original draft:

> Many systems will also have some kind of interaction with the outside world. This could mean clients that use the services over the internet, such as browsers, mobile apps or IoT devices. Typical communication would either be via standard HTTP or WebSocket, but usually not directly to each service. There will usually be a network boundary acting as a perimeter, and a well-controlled communication point that acts as the intermediary between the outside world and the inside world. This is what we call a service gateway, and you can envision it as a Medieval town with a wall around it and a literal gateway that forms the only way in or out. Communication within the walls is more free and direct, but communication to or from the outside world needs to travel through the gateway.